guides/keras_cv/simsiam-with-kerascv.py

```diff
@@ -0,0 +1,604 @@
+    # towards something like 0.0220. The actual number will depend on the layer size.
+    o = tfsim.layers.ActivationStdLoggingLayer(name="proj_std")(x)
+    projector = tf.keras.Model(inputs, o, name="projector")
+    return projector
+
+
+projector = get_projector(input_dim=backbone.output.shape[-1], dim=DIM, num_layers=2)
+projector.summary()
+
+
+"""
+Finally, we must construct the predictor.  The predictor is used in SimSiam, and is a
+simple stack of two MLP layers, containing a bottleneck in the hidden layer.
+"""
+
+
+def get_predictor(input_dim, hidden_dim=512, activation="relu"):
+    inputs = tf.keras.layers.Input(shape=(input_dim,), name="predictor_input")
+    x = inputs
+
+    x = tf.keras.layers.Dense(
+        hidden_dim,
+        use_bias=False,
+        kernel_initializer=tf.keras.initializers.LecunUniform(),
+        name="predictor_layer_0",
+    )(x)
+    x = tf.keras.layers.BatchNormalization(
+        epsilon=1.001e-5, name="batch_normalization_0"
+    )(x)
+    x = tf.keras.layers.Activation(activation, name=f"{activation}_activation_0")(x)
+
+    x = tf.keras.layers.Dense(
+        input_dim,
+        kernel_initializer=tf.keras.initializers.LecunUniform(),
+        name="predictor_output",
+    )(x)
+    # Metric Logging layer. Monitors the std of the layer activations.
+    # Degnerate solutions colapse to 0 while valid solutions will move
+    # towards something like 0.0220. The actual number will depend on the layer size.
+    o = tfsim.layers.ActivationStdLoggingLayer(name="pred_std")(x)
+    predictor = tf.keras.Model(inputs, o, name="predictor")
+    return predictor
+
+
+predictor = get_predictor(input_dim=DIM, hidden_dim=512)
+predictor.summary()
+
+
+"""
+## Training
+
+First, we need to initialize our training model, loss, and optimizer.
+"""
+loss = tfsim.losses.SimSiamLoss(projection_type="cosine_distance", name="simsiam")
+
+contrastive_model = tfsim.models.ContrastiveModel(
+    backbone=backbone,
+    projector=projector,
+    predictor=predictor,  # NOTE: simiam requires predictor model.
+    algorithm="simsiam",
+    name="simsiam",
+)
+lr_decayed_fn = tf.keras.optimizers.schedules.CosineDecay(
+    initial_learning_rate=INIT_LR,
+    decay_steps=PRE_TRAIN_EPOCHS * PRE_TRAIN_STEPS_PER_EPOCH,
+)
+wd_decayed_fn = tf.keras.optimizers.schedules.CosineDecay(
+    initial_learning_rate=WEIGHT_DECAY,
+    decay_steps=PRE_TRAIN_EPOCHS * PRE_TRAIN_STEPS_PER_EPOCH,
+)
+optimizer = tfa.optimizers.SGDW(
+    learning_rate=lr_decayed_fn, weight_decay=wd_decayed_fn, momentum=0.9
+)
+
+"""
+Next we can compile the model the same way you compile any other Keras model.
+"""
+
+contrastive_model.compile(
+    optimizer=optimizer,
+    loss=loss,
+)
+
+"""
+We track the training using several callbacks.
+
+* **EvalCallback** creates an index at the end of each epoch and provides a proxy for the nearest neighbor matching classification using `binary_accuracy`.
```

Comment:
How is the binary accuracy calculated?

Reply:
We can pull from the description in the [module](https://github.com/tensorflow/similarity/blob/master/tensorflow_similarity/classification_metrics/binary_accuracy.py) in tfsim.

```
    Calculates how often the query label matches the derived lookup label.

    Accuracy is technically (TP+TN)/(TP+FP+TN+FN), but here we filter all
    queries above the distance threshold. In the case of binary matching, this
    makes all the TPs and FPs below the distance threshold and all the TNs and
    FNs above the distance threshold.

    As we are only concerned with the matches below the distance threshold, the
    accuracy simplifies to TP/(TP+FP) and is equivalent to the precision with
    respect to the unfiltered queries. However, we also want to consider the
    query coverage at the distance threshold, i.e., the percentage of queries
    that retrun a match, computed as (TP+FP)/(TP+FP+TN+FN). Therefore, we can
    take $ precision \times query_coverage $ to produce a measure that capture
    the precision scaled by the query coverage. This simplifies down to the
    binary accuracy presented here, giving TP/(TP+FP+TN+FN).
```